otlp/traces_direct_json.go

```diff
@@ -0,0 +1,1115 @@
+			return nil, err
+		}
+
+		// If this is an exception event, also collect exception-specific attributes
+		if bytes.Equal(name, []byte("exception")) {
+			exceptionAttrs = msgpAttributesPool.Get().(*msgpAttributes)
+
+			// Extract exception attributes
+			for _, kv := range attributes {
+				kvObj, err := kv.Object()
+				if err != nil {
+					continue
+				}
+
+				var keyStr string
+				var valueObj *fastjson.Value
+
+				kvObj.Visit(func(k []byte, v *fastjson.Value) {
+					switch string(k) {
+					case "key":
+						keyStr = string(v.GetStringBytes())
+					case "value":
+						valueObj = v
+					}
+				})
+
+				if keyStr == "" || valueObj == nil {
+					continue
+				}
+
+				switch keyStr {
+				case "exception.message", "exception.type", "exception.stacktrace":
+					valObj, err := valueObj.Object()
+					if err == nil {
+						valObj.Visit(func(k []byte, v *fastjson.Value) {
+							if string(k) == "stringValue" {
+								valBytes := v.GetStringBytes()
+								exceptionAttrs.addString([]byte(keyStr), valBytes)
+							}
+						})
+					}
+				case "exception.escaped":
+					valObj, err := valueObj.Object()
+					if err == nil {
+						valObj.Visit(func(k []byte, v *fastjson.Value) {
+							if string(k) == "boolValue" {
+								exceptionAttrs.addBool([]byte(keyStr), v.GetBool())
+							}
+						})
+					}
+				}
+			}
+		}
+	}
+
+	// Set event name
+	eventAttr.addString([]byte("name"), name)
+
+	// Calculate duration relative to span start
+	if timeUnixNano > 0 && spanStartTime > 0 {
+		if timeUnixNano >= spanStartTime {
+			timeSinceSpanStart := float64(timeUnixNano-spanStartTime) / float64(time.Millisecond)
+			eventAttr.addFloat64([]byte("meta.time_since_span_start_ms"), timeSinceSpanStart)
+		} else {
+			// Event time is before span start time
+			eventAttr.addFloat64([]byte("meta.time_since_span_start_ms"), float64(0))
+			eventAttr.addBool([]byte("meta.invalid_time_since_span_start"), true)
+		}
+	}
+
+	// Add error status from parent span if applicable
+	if isError {
+		eventAttr.addBool([]byte("error"), true)
+	}
+
+	// Add resource and scope attributes
+	eventAttr.addAttributes(scopeAttrs)
+	eventAttr.addAttributes(resourceAttrs)
+
+	// Set timestamp
+	timestamp := timestampFromUnixNano(timeUnixNano)
+
+	attrBuf, err := eventAttr.finalize()
+	if err != nil {
+		return nil, err
+	}
+	event := EventMsgp{
+		Attributes: attrBuf,
+		SampleRate: sampleRate,
+		Timestamp:  timestamp,
+	}
+	batch.Events = append(batch.Events, event)
+	return exceptionAttrs, nil
+}
+
+// unmarshalSpanLinkJSON parses a Span.Link message from JSON and creates an event
+func unmarshalSpanLinkJSON(
+	ctx context.Context,
+	v *fastjson.Value,
+	traceID,
+	parentSpanID,
+	parentName []byte,
+	parentTimestamp time.Time,
+	resourceAttrs,
+	scopeAttrs *msgpAttributes,
+	sampleRate int32,
+	isError bool,
+	batch *BatchMsgp,
+) error {
+	eventAttr := msgpAttributesPool.Get().(*msgpAttributes)
+
+	// Set trace info
+	if len(traceID) > 0 {
```

Comment:
Are trace_id and trace.parent_id not required fields on span links and span events?

Reply:
Trace id, yes. Parent id, I'm not so sure. If it were missing, it would still be usable in some ways.